django/db/backends/oracle/operations.py

```diff
@@ -315,17 +315,71 @@ def savepoint_create_sql(self, sid):
     def savepoint_rollback_sql(self, sid):
         return "ROLLBACK TO SAVEPOINT " + self.quote_name(sid)
 
+    def _foreign_key_constraints(self, table_name):
+        with self.connection.cursor() as cursor:
+            cursor.execute("""
+                SELECT
+                    user_tables.table_name, rcons.constraint_name, MAX(level)
+                FROM
+                    user_tables
+                JOIN
+                    user_constraints cons
+                    ON (user_tables.table_name = cons.table_name AND cons.constraint_type = ANY('P', 'U'))
+                LEFT JOIN
+                    user_constraints rcons
+                    ON (user_tables.table_name = rcons.table_name AND rcons.constraint_type = 'R')
+                START WITH user_tables.table_name = UPPER(%s)
+                CONNECT BY NOCYCLE PRIOR cons.constraint_name = rcons.r_constraint_name
+                GROUP BY
+                    user_tables.table_name, rcons.constraint_name
+                HAVING user_tables.table_name != UPPER(%s)
+                ORDER BY MAX(level) DESC
+            """, (table_name, table_name))
+            return [
+                (foreign_table, constraint) for foreign_table, constraint, _ in cursor.fetchall()
+            ]
+
     def sql_flush(self, style, tables, sequences, allow_cascade=False):
-        # Return a list of 'TRUNCATE x;', 'TRUNCATE y;',
-        # 'TRUNCATE z;'... style SQL statements
         if tables:
-            # Oracle does support TRUNCATE, but it seems to get us into
-            # FK referential trouble, whereas DELETE FROM table works.
-            sql = ['%s %s %s;' % (
-                style.SQL_KEYWORD('DELETE'),
-                style.SQL_KEYWORD('FROM'),
-                style.SQL_FIELD(self.quote_name(table))
-            ) for table in tables]
+            truncated_tables = {table.upper() for table in tables}
+            constraints = set()
+            if allow_cascade:
+                # Oracle's TRUNCATE CASCADE only works with ON DELETE CASACADE
+                # foreign keys which Django doesn't define. Emulate the
+                # PostgreSQL behavior which truncates all dependent tables by
+                # manually retrieving all foreign key constraints and resolving
+                # dependencies.
+                for table in tables:
+                    for foreign_table, constraint in self._foreign_key_constraints(table):
```

Comment:
I don't think it's necessary. This approach was truncating and switching constraints on/off multiple times if the table appeared in the dependency cycle more than once. The `backends.tests.LongNameTest.test_sequence_name_length_limits_flush` test is failing when calling sqlflush without `allow_cascade=True`.